optuna/integration/mlflow.py

```diff
@@ -13,8 +14,79 @@
     import mlflow
 
 
+class MLFlowIntegrator(object):
+    def __init__(
+        self,
+        tracking_uri: Optional[str] = None,
+        metric_name: str = "value",
+        nest_trials: bool = False,
+        tag_study_user_attrs: bool = False,
+    ) -> None:
+
+        _imports.check()
+
+        self._tracking_uri = tracking_uri
+        self._metric_name = metric_name
+        self._nest_trials = nest_trials
+        self._tag_study_user_attrs = tag_study_user_attrs
+
+    def set_experiment(self, study: optuna.study.Study) -> None:
+        # This sets the tracking_uri for MLflow.
+        if self._tracking_uri is not None:
+            mlflow.set_tracking_uri(self._tracking_uri)
+
+        # This sets the experiment of MLflow.
+        mlflow.set_experiment(study.study_name)
+
+    def calculate_tags(self, trial: optuna.trial.BaseTrial, study: optuna.study.Study) -> None:
+
+        tags: Dict[str, str] = {}
+        tags["number"] = str(trial.number)
+        tags["datetime_start"] = str(trial.datetime_start)
+        if isinstance(trial, optuna.trial.FrozenTrial):
```

Comment:
It is quite possible that I got something wrong. I have not been looking into the internals of Optuna before, but the property `datetime_complete` is not available on other types of trial after my observation.

Reply:
Thank you for your clarification. You are right! Related to the comment, can I ask why the trial's type is relaxed from `FrozenTrial` to `baseTrial` in the argument of `set_tags`?

Reply:
`set_tags` is called from the Callback and the decorator.
The callback is called after the result of the optimization function has been stored by Optuna and the `Trial` has been transformed to a `FrozenTrial`. However in the context of the decorator `set_tag` is called before returning back to the Study and hence it is still a `Trial`. I thought that the most proper way to handle this difference for this two cases using this one instance check. As always I am open for other ideas.


Just as information MLFlow also tracks the duration of an trial/run. However when using the Callback functionalty this value is not correct since for MLFlow the run only starts in the callback

Reply:
In the Callback we are working with a `FrozenTrial` (executed after the optimization function call has been handled by the study.
In the Decorator we did not return the to Study yet and hence still have a `Trial`.
That's why I had to relax to `BaseTrial` and introduce the instance check. 

If you have any better idea how to approach this, I am always open.

Just as a side note: MLFlow tracks to start and end of a run (=trial) as well. However when using the Callback functionality this information is not accurate since the run is only started after the trial has ended.

Reply:
Thank you for your clarification. So `@track_in_mlflow` cannot save`TrialState` and `datetime_complete` as in the tag attributes.

Reply:
I think one way to resolve it is to resume the run; `track_in_mlflow` provides only `with` block to enable users to call mlflow's logging function and stores the `run_id`, then we resume the run by using `run_id` in the callback to perform mlflow's logging for `FrozenTrial`.

https://github.com/nzw0301/optuna/commit/67943f1f2fd0f977043aaef5f96cf86314103e97

I believe there is room to improve the API, but what do you think of this?

Reply:
I like it. It is less a break and more an extension of the current behaviour than my proposition
Two points I think should be discussed
- What happens when somebody does not add the Callback but sets the track_in_mlflow -> Optuna does not log anything (I can life with that)
- Multi-threaded behavior: What happens if the `nb_jobs` is set in the `optimize`-call. I know it is deprecated but still available until v4.

I am not that sure if if an attribute is the right place to store the run-id for me it looks dangerous for potential multi-threading and exception handling. My guts tell me it would be better to save it on the trial. Could `system_attrs` be used for that?

Reply:
Thanks for fruitful discussions @nzw0301 and @lucafurrer. I agree with @nzw0301's idea. For the concerns by @lucafurrer, I think
- If users set `track_in_mlflow` but the callback, Optuna should not log anything. It is sufficient to document the behavior enough.
- If the `n_jobs` argument is set, `MLFlowIntegrator` should raise an ValueError until v4. It is also sufficient to document the behavior.

> I am not that sure if if an attribute is the right place to store the run-id for me it looks dangerous for potential multi-threading and exception handling.

If we decide not to support the multi-threading with this MLFlow integration, we don't have to consider this problem, right? And, I'm not sure about the "exception handling". Could you elaborate a little more about it?

Reply:
I have not thought all cases through, but since the proposed flow is
1. Set id in callback class in decorator
2. Read out id when in callback function
3. Reset id in callback class
Whenever something happens between 1 and 3 the id would not be reset. But since we would overwrite it in 1 again there would not be much harm to be done. 

But I still would prefer to store the information on the trial than in a "persistent" variable.

Reply:
We @nzw0301 approach by saving the run_id in a persistent variable in the class we have the following flow
1. Set run_id
2. Read run_id
Unless 1 is run again, the id is never reset. So if out of some error 1 is not set properly for the next trial, we overwrite the existing trial.

Other problematic pattern could be
```python
mlflc = MLFlowCallback(...)

@mlflc.track_in_mlflow()
def objective1(trial):
   mlflow.log_metric('something', 1)
   ...
   return x

def objective2(trial)
   ...
   return x*x

study1 = optuna.create_study(study_name="my_study")
study1.optimize(objective1, n_trials=10, callbacks=[mlflc])

study2 = optuna.create_study(study_name="my_study")
study2.optimize(objective2, n_trials=10, callbacks=[mlflc])
```
This might be far from best practice, but nothing prevents somebody to do this and the seconde study will log everything into the last run from the previous study.

I think it is possible to create other problematic scenarios with saving the current run into a variable of the class. This problems could be avoided if the run_id is not stored in MLFlowCallback but on the trial p.ex in `system_attrs` or `user_attrs`

Reply:
Thank you for your informative comments! As another scenario, we might want to resume optimization of the study like your example provided just before the comment of this comment with `optuna.study.load_study`. So personally, a re-usable callback sounds fine for some cases.

Reply:
Ok this should be fine with any of the implementations. I was also thinking about the fact that in any cases it might be useful to have a reference to the MLFlow-Run-ID stored in Optuna to allow jumping from one to an other. Still not sure if it should be `system_attrs` or `user_attrs`